### PR TITLE
Add error notification for tile count limit exceeded

### LINKED ISSUE #211: Mapbox tile count limit error notification not being shown on (some) devices
This has been observed on Android version 8

### How to replicate

1. Open the Kujaku Sample app
2. Enter a name in the **Offline Map Download** field
3. Click on the **Start Offline Download** button

The download will start and a notification will be shown, however at around 2.34%, the download stops, on checking the logs, there's a Mapbox Tile count limit error

The offline regions page also becomes unresponsive -> When openned, nothing is displayed and the usual toast to inform user that there are no offline regions does not show up. This probably means the Offline manager has hang and is not responding to calls for listing offline regions.

- [x] Further investigation is also required to make sure the host application can recover from this error
- [x] The offline maps download service should be exited in such a case
- [x] An appropriate error notification should be shown
- [x] The host application should be notified that this has happened

### PR DESCRIPTION
**Pull request by ekigamba**
Fixes #211 

See #211 for more

### How to test

1. Enter a download name in the first screen after openning the app. Do not enter any coordinates for it to download the default map
2. The map download tile count will be reached at around 2.34%. At this point, the error notification should appear

### CODE REVIEW

File: library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java [hunk 1]
```diff
@@ -261,7 +261,7 @@ public void mapboxTileLimitExceededShouldCreateValidBroadcast() throws Interrupt
         latch.await();
 
         Intent intent = (Intent) resultsToCheck.get(0);
-        assertBroadcastResults(intent, MapboxOfflineDownloaderService.SERVICE_ACTION_RESULT.FAILED, mapName, "MapBox Tile Count limit exceeded : 60000 while Downloading " + mapName, MapboxOfflineDownloaderService.SERVICE_ACTION.DOWNLOAD_MAP);
+        assertBroadcastResults(intent, MapboxOfflineDownloaderService.SERVICE_ACTION_RESULT.FAILED, mapName, "MapBox Tile Count limit exceeded 60000 while Downloading " + mapName, MapboxOfflineDownloaderService.SERVICE_ACTION.DOWNLOAD_MAP);
```
Comment (vincent-karuri) [inline library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java]: I think the error string can be replaced with the string resource `error_mapbox_tile_count_limit`?

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java [hunk 2]
```diff
@@ -110,9 +111,12 @@
     private long currentMapDownloadId;
 
     private DownloadProgressNotification downloadProgressNotification;
-    public static final int PROGRESS_NOTIFICATION_ID = 85;
+    public static final int PROGRESS_NOTIFICATION_ID = 80;
     public static final int REQUEST_ID_STOP_MAP_DOWNLOAD = 1;
-    public int LAST_DOWNLOAD_COMPLETE_NOTIFICATION_ID = 87;
+
+    // Should persist across service instances but within the same app session
+    public static int LAST_DOWNLOAD_COMPLETE_NOTIFICATION_ID = 180;
```
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Is the assumption here that the `DownloadCompleteNotification` will not be shown more that 100 times within the app's life-time? 

Is there an advantage to this approach than using say `Math.random() * int` to generate the notification ids?

File: library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java (hunk 1 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java]: Cool

**Comment by vincent-karuri** [comment]
screenshot_20181210-105219

On the Samsung J5 mobile phone, I get two notifications, should this happen?

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 2 above)
Comment (vincent-karuri) [inline library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Is the assumption here that the DownloadCompleteNotification will not be shown more that 100 times within the app's life-time?

**Comment by ekigamba** [comment]
> On the Samsung J5 mobile phone, I get two notifications, should this happen?

Yea, one notification is produced by the library while the other is produced by the sample app. The notification produced by the sample app is generic shows all kind of errors. Do you want me to remove the mapbox tile count limit error notification shown by the sample app?

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 2 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: > Is the assumption here that the DownloadCompleteNotification will not be shown more that 100 times within the app's life-time?

Within the application lifecycle not the life-time

File: library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java (hunk 1 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java]: Not sure I understand. I think it's okay for tests to fail after changes, right?

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: Yea, that's what I meant.

So if the application is not killed for a while, will this ever be an issue? Could the 100 limit ever be crossed?

File: library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java (hunk 1 above)
Comment (ekigamba) [reply library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java]: This error string is already in `strings.xml`, if we use the string resource then the tests will not fail when this is changed in the actual application code

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 2 above)
Comment (ekigamba) [reply library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: **Could a host application successfully download 100 maps within one application life-cycle?** Realistically not, but possibly yes.

I'll increase the gaps to 1000. Is that ok?

File: library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java (hunk 1 above)
Comment (vincent-karuri) [reply library/src/test/java/io/ona/kujaku/services/MapboxOfflineDownloaderServiceTest.java]: okay.

File: library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java (hunk 2 above)
Comment (vincent-karuri) [reply library/src/main/java/io/ona/kujaku/services/MapboxOfflineDownloaderService.java]: That's fine.

**Comment by vincent-karuri** [comment]
Hadn't seen this before but could you close/remove the progress downloading notification when the error downloading notification shows up.

The two appearing together can be confusing.

**Pull request merged**